torch/csrc/distributed/rpc/py_rref.cpp

```diff
@@ -54,18 +104,17 @@ PyRRef::PyRRef(c10::intrusive_ptr<RRef> rref) : rref_(std::move(rref)) {
   TORCH_CHECK(rref_, "PyRRef must not wrap nullptr");
 }
 
-PyRRef::PyRRef(const py::object& value)
-    : PyRRef([&value]() {
-        jit::InferredType type_inferred = jit::tryToInferType(value);
-        TypePtr elem_type = nullptr;
-        if (type_inferred.success()) {
-          // If we could infer the type from the pyobject, we create
-          // the RRef with the IValue of that type.
-          elem_type = type_inferred.type();
-        } else {
-          // Otherwise it's a pure pyobject, create the RRef
-          // that holds an IValue of an pyobject
-          elem_type = PyObjectType::get();
+PyRRef::PyRRef(const py::object& value, const py::object& type_hint_py)
+    : PyRRef([&value, &type_hint_py]() {
+        TypePtr type_hint_ptr = tryTypePyObjToTypePtr(type_hint_py);
+        TypePtr elem_type = decidePyObjJitType(value, type_hint_ptr);
+        if (type_hint_ptr != nullptr) {
+          TORCH_CHECK(
+              elem_type->isSubtypeOf(type_hint_ptr),
+              "The specificied type hint is ",
+              type_hint_ptr->python_str(),
+              "The decided type is ",
```

Comment:
that's kind of confusing to user, user shouldn't know what a specified type or a decided type means. We should just error out and saying: 
type_hint_ptr->python_str() is not a subtype of elem_type->python_str()

Reply:
also, this seems to be already checked in `decidePyObjJitType` from what i see. 

For the API usage of this `type_hint`, We should only allow user to specify the type_hint when the type_hint is a jit class type/interface type. That means if user pass a type that is not a class type, we should guard against it, because that's also what `tryTypePyObjToTypePtr` is trying to do. But the current state of this PR allows user to possibly pass types other than class types (like Int, Tensor, etc.) and will errors with `Please script the type to register it before using it.` we should instead tell the user sth like "type_hint only accepts TorchScript class/interface"

Reply:
Right, I am allowing any type hint. Let me narrow it down to ScriptModule. 
Need to add test case for ScriptClass.